Enable tests/crypto/enclave on Windows

CodeMonkeyLeet opened the pull request:
- Enable tests/crypto/enclave on Windows

    - Enable building tests/crypto/enclave folder on Windows.

    - Remove f_readv in crypto.edl, which cannot compile on Windows.

    - Refactor SYS_readv syscall hook in crypto/enclave/enc/enc.c to
      call f_read instead of f_readv since it was already using a
      synchronous single-buffer design for the ocall.

    - Update f_* methods in crypto/enclave/host/host/c to open and
      read files as binary on Windows.

    - Fix tests/crypto/enclave/CMakeLists.txt to use add_enclave_test
      instead of add_test.

    - Fix add_enclave_test.cmake to support nested test paths when
      using the ADD_WINDOWS_ENCLAVE_TESTS flag.

- Refactor tests/crypto/data build for Windows

    - Move key and certificate generation actions into dedicated bash script
      in tests/crypto/data so that it can execute in the git bash environment
      during Windows build.

    - Build tests/crypto/data as a separate CMake step to remove implicit
      dependency of enclave tests on build of the host tests.

    - Standardize naming conventions for crypto test data.

- Refactor tests/crypto_crls_cert_chains/data build for Windows

    - Move key and certificate generation actions into dedicated bash script
      in tests/crypto_crls_cert_chains/data so that it can execute in the git
      bash environment during Windows build.

    - Build tests/crypto/data as a separate CMake step to remove implicit
      dependency of enclave tests on build of the host tests.

    - Standardize naming conventions for crypto_crls_cert_chains test data.

CodeMonkeyLeet commented on cmake/add_enclave_test.cmake:
```diff
@@ -5,61 +5,71 @@
-## DESCRIPTION : For ADD_WINDOWS_ENCLAVE_TESTS enabled function will copy signed 
-##			enclave file from Linux build location to windows build location 
-##			after checking if both host and enclave file exists at specified 
+## ENC_FILE 	: Signed/UnSigned enclave file name.
```
Fixed

CodeMonkeyLeet commented on cmake/add_enclave_test.cmake:
```diff
@@ -5,61 +5,71 @@
+        # a corresponding subpath to the linux binaries.
+        set(OE_TESTS_PATH_NAME "tests/")
+        string(LENGTH ${OE_TESTS_PATH_NAME} OE_TESTS_PATH_NAME_LENGTH)
+        string(FIND ${CMAKE_CURRENT_SOURCE_DIR} ${OE_TESTS_PATH_NAME} TEST_DIR_INDEX)
```
Good catch, fixed

CodeMonkeyLeet commented on tests/crypto/host/CMakeLists.txt:
```diff
@@ -21,134 +20,7 @@ add_executable(hostcrypto
-    ../utils.c)
-endif()
-
 target_link_libraries(hostcrypto oehost)
```
Fixed. Also updated tests/crypto/enclave/host/CMakeLists.txt

CodeMonkeyLeet commented on tests/crypto_crls_cert_chains/host/CMakeLists.txt:
```diff
@@ -4,88 +4,5 @@
 oeedl_file(../common/crypto_crls_cert_chains.edl host gen)
 
 add_executable(crypto-extra_host host.cpp ${gen})
```
Fixed

anitagov commented on scripts/ansible/roles/windows/az-dcap-client/vars/windows.yml:
```diff
@@ -15,3 +15,22 @@ lc_driver:
+  - "C:\\openenclave\\prereqs\\nuget\\Microsoft.Azure.DCAP.Client"
+  - "C:\\openenclave\\prereqs\\nuget\\openssl"
+  - "C:\\openenclave\\prereqs\\nuget\\zlib"
+
```
Why do we need openssl? For Windows, we have been using openssl from the Git Bash directory.
Why is zlib needed?

CodeMonkeyLeet replied:
None of these are relevant. The bad rebase has been dropped.

anitagov replied:
Great. I later saw that this was already in the master.

anitagov commented on scripts/ansible/roles/windows/az-dcap-client/vars/windows.yml:
```diff
@@ -15,3 +15,22 @@ lc_driver:
+
+validation_binaries:
+  - "C:\\openenclave\\prereqs\\nuget\\curl\\build\\native\\lib\\v110\\x64\\Release\\static\\libcurl.lib"
+  - "C:\\openenclave\\prereqs\\nuget\\DCAP_Components\\build\\lib\\native\\Libraries\\sgx_dcap_ql.lib"
```
Version number is missing.

anitagov commented on scripts/ansible/roles/windows/az-dcap-client/vars/windows.yml:
```diff
@@ -15,3 +15,22 @@ lc_driver:
+validation_binaries:
+  - "C:\\openenclave\\prereqs\\nuget\\curl\\build\\native\\lib\\v110\\x64\\Release\\static\\libcurl.lib"
+  - "C:\\openenclave\\prereqs\\nuget\\DCAP_Components\\build\\lib\\native\\Libraries\\sgx_dcap_ql.lib"
+  - "C:\\openenclave\\prereqs\\nuget\\EnclaveCommonAPI\\lib\\native\\x64-Release\\sgx_enclave_common.lib"
```
Version number is missing

anitagov commented on scripts/ansible/roles/windows/az-dcap-client/vars/windows.yml:
```diff
@@ -15,3 +15,22 @@ lc_driver:
+  - "C:\\openenclave\\prereqs\\nuget\\DCAP_Components\\build\\lib\\native\\Libraries\\sgx_dcap_ql.lib"
+  - "C:\\openenclave\\prereqs\\nuget\\EnclaveCommonAPI\\lib\\native\\x64-Release\\sgx_enclave_common.lib"
+  - "C:\\openenclave\\prereqs\\nuget\\libssh2\\build\\native\\lib\\v110\\x64\\Release\\static\\cdecl\\libssh2.lib"
+  - "C:\\openenclave\\prereqs\\nuget\\Microsoft.Azure.DCAP.Client\\lib\\release\\dcap_quoteprov.dll"
```
Version number is missing

CodeMonkeyLeet commented:
@anitagov I took a look at that last night after our conversation and the failures are existing ones, triggered whenever tests/crypto/enclave attempts to printf from inside the enclave. It would have still spewed the error, just not with the diagnostic message. 

It really should have been using the corelibc version as `common` code does, but the syscall check is essentially noise since it's allowed to fallthrough the syscall handler to the explicit redirect to oe_host_printf in the enclave.

Removed the use of OE_RAISE and replaced it with a trace if verbose.

CodeMonkeyLeet commented:
bors r+

andschwa commented on tests/crypto/data/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+    message(FATAL_ERROR "Unknown OS. Only supported OSes are Linux and Windows")
+endif()
+
+add_custom_target(
```
Are you sure you want this instead of `add_custom_command`? The explicit listing of `BYPRODUCTS` makes me think you mean `OUTPUT`...

CodeMonkeyLeet replied:
The choice of using a custom target is intended so that the collection of collateral generated is a single node in the cmake graph for the host/enclave folders to take a dependency on.

The listing of byproducts is really just to enable potential optimization for ninja, they're not explicit targets.

andschwa commented on tests/crypto/data/make-test-certs:
Oh goodness...so much fun...

andschwa commented on tests/crypto/host/CMakeLists.txt:
```diff
@@ -21,134 +20,8 @@ add_executable(hostcrypto
-
-# ========================= TestCRL & TestASN1 ================================
-
-add_custom_command(TARGET hostcrypto
```
Just curious, what was the reasoning of replacing this with a Bash script?

CodeMonkeyLeet replied:
Many of the `COMMAND` lines are linux commands (`sleep`, `awk`, `tail`, etc.) which cmake runs in CMD and just fail. Encapsulating them in a bash script allows a clearer specification of the execution environment instead of trying to branch individual commands for cross-platform compat at the cmake level.
